### DISCUSSION TITLE
Howto run a worker thread in background

### DISCUSSION THREAD

**Discussion by hoderlump** [zauberzeug/nicegui#836] [Q&A] [answered]
### Question

Hi!

I would like to run a thread in the background of the GUI to do some stuff in realtime.
So I want to initialise an object of class and run some functions from it in a background thread.
I want to do this by pressing a button.

btw: I learned that the GUI should always run in the main thread. Is that true also for NiceGUI?

So, this is what I do:
```python
import threading
from nicegui import ui

class MyClass:
    """A simple example class"""
    i = 12345

    def f(self):
        print('hello world')

def start_sm():
    # create statemachine instance:
    sm = MyClass()

    sm_thread = threading.Thread(target=run_sm, args=(sm,))
    sm_thread.daemon = True
    sm_thread.start()

def run_sm(sm):
    t = threading.currentThread()
    sm.f()

# ui elements:
ui.button('Start my background worker!', on_click=start_sm())

ui.run()
```

My problem is: The class is instantiated and executed in a thread two times (!) before the GUI opens. I do not know why this happens. This is the output:
```
hello world
hello world
NiceGUI ready to go on http://0.0.0.0:8080
```

Why is the function start_sm() even executed before the GUI is built up?
After I press the button nothing happens...

How do you do execute background threads in NiceGUI? 

Thanks for your help. I did not find any examples on the web but I must say I am new to GUI programming. If my way to do is is completely wrong, please tell me... :)
Either way, this program needs threads to perform properly- the background class i quite huge and runs from the command line quite well. I want to start and control it using a GUI.

Thanks

**Comment by rodja** [comment]
Read https://github.com/zauberzeug/nicegui/issues/794 to avoid the double-execution. NiceGUI runs on the main thread and uses the asyncio event-loop. As a concurrency model, async/await is my absolute favourite because it allows you to write linear code where you would otherwise create lot of problems with threads. May I suggest the excellent explanations about [async/await from FastAPI](https://fastapi.tiangolo.com/async/?h=async)? But if you have existing code which is not easily rewritten, you can of course also mix the two concurrency models.

**Comment by hoderlump** [comment]
Thanks for you answer!
I see now how the async/await mechanic works.
Even without starting an exta threas, non of the solutions in https://github.com/zauberzeug/nicegui/issues/794 does help here:

- I need to instantiate an object of a class only once
- I need this object to be accessible or to be in the scope than the GUI so that the GUI can interact with it

Can this be done at all with the way NiceGUI works?

**Comment by rodja** [comment]
Wrapping your code in a startup method as suggested in #794 works for me:

```py
from nicegui import app, ui

class MyClass:
    """A simple example class"""
    i = 12345

    def f(self):
        print('hello world')


def startup():
    def start_sm():
        # create statemachine instance:
        sm = MyClass()

        sm_thread = threading.Thread(target=run_sm, args=(sm,))
        sm_thread.daemon = True
        sm_thread.start()

    def run_sm(sm):
        t = threading.currentThread()
        sm.f()

    # ui elements:
    ui.button('Start my background worker!', on_click=start_sm)

app.on_startup(startup)

ui.run()
```

The only other thing which caused a problem was the execution of `start_sm` when the button had been instantiated. Instead, the function itself must be set as on-click-handler.

**Comment by hoderlump** [comment, marked as answer]
I think I have a solution. With the following code I can instantiate an object of a class only once. It does not use threads but maybe this is enough for the moment. Maybe my initial question was not specific enough...

This class instance can also be accessed from GUI elements by making the class instance global using the `global` keyword. This way I can execute functions from the class and access its variables.
By using `lambda` an the cryptic `setattr` function I could prevent NiceGUI from executing these functions while drawing the ui. Very odd, but it works well. :)

```python
from nicegui import app, ui

class MyClass:

    def f(self):
        print(f'class function executed! Class variable is {str(self.i)}.')

    def __init__(self):
        self.i = 0
        print(f'class instantiated! Class variable is {str(self.i)}.')


def instantiate():
    global sm
    sm = MyClass()

    ui.button('Run my class function!', on_click=lambda: sm.f())
    ui.button('Increment class variable', on_click=lambda: setattr(sm, "i", sm.i+1))

    container.remove(initialise_button)

# ui elements:
container = ui.column()
with container:
    initialise_button = ui.button('Instantiated my class!', on_click=lambda: initialise())

ui.button('shutdown', on_click=app.shutdown)

ui.run(reload=False)
```

Maybe I don't need threads- an instance of the class which does what it should do could be enough here.
I have to check if the UI stays responsive while the class does its stuff...

By including the following lines before `ui.run(reload=False)` the class can be instantiated during the startup of the ui:
```python
if __name__ == "__main__":
    instantiate()
```
In case of `reload=True`, this works for automatic instantiation:
```python
if __name__ == "__mp_main__":
    instantiate()

ui.run(reload=True)
```
It would be good if this `reload `option would be documented somewhere. It can be very confusing. The backgroud thread it creates needs to be killed if I run or start to debug the code again manually from an IDE.

@rodja, thanks for your quick help here and the nice framework! :)

**Reply by rodja** [reply]
> By using lambda an the cryptic setattr function I could prevent NiceGUI from executing these functions while drawing the ui.

Interesting solution. Lambda statements can only call functions, not do assignments. But with `setattr` you found a work around.

> It would be good if this reload option would be documented somewhere.

We now have a separate documentation page for `ui.run`: [run_documentation.py](https://github.com/zauberzeug/nicegui/blob/main/website/more_documentation/run_documentation.py) is shown as https://nicegui.io/documentation/run. Would you like to add some explanations in a pull request. Much appreciated.
